runtime/intern.c

```diff
@@ -239,446 +284,395 @@ static void readfloats(struct caml_intern_state* s,
+            caml_named_value ("Debugger.function_placeholder");
+          if (function_placeholder != NULL) {
+            /* Use the code pointer from the "placeholder" function */
+            v = (value) Code_val(*function_placeholder);
+          } else {
             intern_cleanup(s);
-            caml_failwith("input_value: ill-formed message");
+            intern_bad_code_pointer(digest);
           }
         }
+        break;
+      case CODE_INFIXPOINTER:
+        ofs = read32u(s);
+        /* Read a value to *dest, then offset *dest by ofs */
+        PushItem(s);
+        sp->dest = dest;
+        sp->op = OShift;
+        sp->arg = ofs;
+        ReadItems(s, dest, 1);
+        continue;  /* with next iteration of main loop, skipping *dest = v */
+      case CODE_CUSTOM:
+      case CODE_CUSTOM_LEN:
+      case CODE_CUSTOM_FIXED: {
+        uintnat expected_size, temp_size;
+        ops = caml_find_custom_operations((char *) s->intern_src);
+        if (ops == NULL) {
+          intern_cleanup(s);
+          caml_failwith("input_value: unknown custom block identifier");
+        }
+        if (code == CODE_CUSTOM_FIXED && ops->fixed_length == NULL) {
+          intern_cleanup(s);
+          caml_failwith("input_value: expected a fixed-size custom block");
+        }
+        while (*s->intern_src++ != 0) /*nothing*/;  /*skip identifier*/
+#ifdef ARCH_SIXTYFOUR
+        if (code == CODE_CUSTOM_FIXED) {
+          expected_size = ops->fixed_length->bsize_64;
+        } else {
+          s->intern_src += 4;
+          expected_size = read64u(s);
+        }
+#else
+        if (code == CODE_CUSTOM_FIXED) {
+          expected_size = ops->fixed_length->bsize_32;
+        } else {
+          expected_size = read32u(s);
+          s->intern_src += 8;
+        }
+#endif
+        temp_size = 1 + (expected_size + sizeof(value) - 1) / sizeof(value);
+        v = intern_alloc_obj(s, d, temp_size, Custom_tag);
+        Custom_ops_val(v) = ops;
+        size = ops->deserialize(Data_custom_val(v));
+        if (size != expected_size) {
+          intern_cleanup(s);
+          caml_failwith(
+            "input_value: incorrect length of serialized custom block");
+        }
+        if (s->intern_obj_table != NULL)
+          s->intern_obj_table[s->obj_counter++] = v;
+        if (ops->finalize != NULL && Is_young(v)) {
+          /* Remember that the block has a finalizer. */
+          add_to_custom_table (&d->minor_tables->custom, v, 0, 1);
+        }
+        break;
       }
-      /* end of case OReadItems */
-      if (first) {
-        result = v;
-        first = 0;
-      } else {
-        Store_field(dest, curr_field, v);
+      default:
+        intern_cleanup(s);
+        caml_failwith("input_value: ill-formed message");
       }
-      break;
-    default:
-      CAMLassert(0);
     }
   }
-  stack_free(&S);
-  s->stack = NULL;
-  CAMLreturn(result);
+  /* end of case OReadItems */
+  *dest = v;
```

Comment:
Just to make sure I understand: the direct assignment `*dest = v;` is safe because
- either `dest` points within the minor heap for the current domain;
- or `dest` points within a major-heap block freshly allocated by `caml_shared_try_alloc` but not yet visible to the GC, and if `v` is a block it's in the major heap too, so no back-reference is created.

Reply:
That is correct. I'll add a comment in the code. And using `caml_modify` would be incorrect since `*dest` is uninitialised memory, and the deletion barrier in `caml_modify` shouldn't be allowed to mark the old value.